### DISCUSSION TITLE
tailwind 2.0 purge not working

### DISCUSSION THREAD

**Discussion by tan-ahmed** [tailwindlabs/tailwindcss#2963] [Help] [answered]
### I am using next.js, when I push to vercel, it strips everything out. any ideas?

**_tailwind.config.js_**
```
module.exports = {
  purge: ["./pages/**/*.{js,ts,jsx,tsx}", "./components/**/*.{js,ts,jsx,tsx}"],
}
```

**_postcss.config.js_**
```
module.exports = {
  plugins: ["tailwindcss", "postcss-nested", "autoprefixer"],
};
```

**_tailwind.css_**
```
@tailwind base;
@tailwind components;
@tailwind utilities;
```

**Comment by MartinHughes-BPC** [comment, marked as answer]
Hi @tan-ahmed 

I tried using a similar syntax for specifying the paths as you used, and it failed on my build - most of the classes were stripped! I will say I'm not particularly familiar with glob syntax, coming from the Windows world, but if I do something equivalent to the below it works flawlessly.

```
purge: [
"./pages/**/*.js",
"./pages/**/*.ts",
"./pages/**/*.jsx",
"./pages/**/*.tsx",
"./components/**/*.js",
"./components/**/*.ts",
"./components/**/*.jsx",
"./components/**/*.tsx"
]
```

Cheers,
Martin.

**Reply by harrygr** [reply]
This worked for me. If anyone knows why the recommended glob pattern seems to not work correctly I'd be interested.

In my case it seemed some styles survived the purge but other were incorrectly purged. It's not even like the ones being purged were dynamic or anything.

**Reply by BrandonClapp** [reply]
Same issue here. Some styles were being stripped from my production build. Changing the glob to what Martin mentioned above fixed it for me.

Not entirely sure why.

**Reply by rjlg** [reply]
I had the same issue. This resolved it. Thanks @MartinHughes-BPC

**Reply by dieduro** [reply]
Wow! I had been stuck with this issue for quite some time without finding the right fix or resaon for it. Thank you very much for sharing this tip. Saved Me!

**Reply by mawhtin** [reply]
This worked for me as well. It's August now. Why isn't the typical {} pattern working?

**Comment by tan-ahmed** [comment]
Hi @MartinHughes-BPC 
I've tried both below, and it's still running into the same problem. Any ideas?

**_tailwind.config.js_**

```
module.exports = {
  purge: ["./pages/**/*.js", "./components/**/*.js"],
}
```

```
module.exports = {
  purge: {
   layers: ["components", "utilities"],
   content: ["./pages/**/*.js", "./components/**/*.js"],
  }
```

**Reply by MartinHughes-BPC** [reply]
Hmm. Not 100% sure. I've only played around with NextJS in the past. However, I followed the instructions here: https://tailwindcss.com/docs/guides/nextjs to create a new Next app with Tailwind. When I run either development or production builds, it works as expected. I don't have an account so can't push anything to Vercel to test that. Is it worth trying to create a new app just for testing and pushing that to Vercel to see if you get the same problem and/or comparing what's different to your app if it does work?

Also wondering if it could be a postcss version issue?

Hope that gives you some ideas to explore....

Cheers,
Martin.

**Reply by tan-ahmed** [reply]
hi @MartinHughes-BPC I created a new git repository, a fresh new installation of tailwind 2.0 with nextjs. Turns out the problem is to do with git itself...
source / solution: https://stackoverflow.com/questions/52950/how-to-make-git-ignore-changes-in-case

many thanks for your help, it works now as expected.

**Comment by simonswiss** [comment]
Hey @tan-ahmed 👋

Are you able to share the code (or a minimal reproduction of the issue) somewhere on GitHub?

Are you by any chance creating dynamic classes based on a prop, like below? That won't work with purge, as it needs to find the complete classname string (in plain text, not evaluated code) to keep it in the output:

```html
<button class={`bg-${props.color}-600 ...`}>click me</button>
```

What happens if you add `enabled:true` to your purge object? Are the classes also stripped out in your local dev?

```diff
module.exports = {
  purge: {
+  enabled: true,
   layers: ["components", "utilities"],
   content: ["./pages/**/*.js", "./components/**/*.js"],
  }
```

**Reply by lmiller1990** [reply]
Adding `enabled: true` worked for me. My project only has 2 files - `style.css` which contains:

```css
/* ./your-css-folder/styles.css */
@tailwind base;
@tailwind components;
@tailwind utilities;
```

and `index.html`.

**Reply by TheCeloReis** [reply]
You saved my life, thanks

**Reply by huzaifarif** [reply]
I think `enabled:true` is not required instead use `NODE_ENV=production` and it will be enabled by default. Also having `enabled: true` without any checks will affect also dev not just prod

**Comment by Johnrobmiller** [comment]
For anybody who is struggling with this despite having tried the solutions above:

The cause of this bug might not have anything to do with any configurations regarding tailwind.config, or at least it didn't for me.  I struggled with this problem for a couple hours just now, and I was able to identify the cause and implement a solution.

For me, the problem was caused by the way I was dynamically using classes in my project.  PurgeCss works by executing a RegEx command on your source code to search for unused classes/styles.  Long story short, it "purges" any styles that aren't being used a given page so that they don't have to load.  Unfortunately, if you are dynamically applying styles and utilities to your elements like I was, then the timing of PurgeCss's RegEx search is slightly awkward...  The utility classes that are being incorrectly "purged" are likely the ones that are dynamically created and never directly referenced in your source code.  For example, in my app, the utility "text-b" is incorrectly getting purged while "text-r" is working correctly.  It is likely no coincidence that that "text-r" is being directly used as a class, while "text-b" is only called dynamically using the variable "textColor" as seen in this code: ```<div className={`some-utility-class ${textColor} some-utility-class`}>```.  Therefore, when the PurgeCss regex searcher looks for classes, it never is able to find "text-b" and erroneously assumes that I'm not using it.

With all this said, the tailwind documentation actually provides a clean solution to this.  I won't repeat what it is because the specific doc in question is worth reading in general anyways, so I'll just link it here: https://tailwindcss.com/docs/optimizing-for-production

**Reply by harrygr** [reply]
You can still dynamically generate your classes. The regex used is deliberately naive. As long as a class appears somewhere in your source code it won't be purged. You just have to make sure the class appears in its entirety.

See https://regex101.com/r/ACVBNg/1 for an example.

The important thing is not to try and construct a class by building from substrings. E.g.

```tsx
const bgColor = 'white'
return <div className={`bg-${bgColor}`} />
```

As `bg-white` never appears in its entirety it will be purged.

In my case this was not the problem. It seemed to be a problem with the glob selector

**Comment by deschantkn** [comment]
If you are like and moved all your nextjs code under a `./src` directory remember to also update the purge paths in your tailwind config file!

**Reply by magnuslundstrom** [reply]
This worked for me, thanks a-lot @deschantkn 

So my tailwind.config.js now looks like this:
```
module.exports = {
  purge: [
    './src/pages/**/*.{js,ts,jsx,tsx}',
    './src/components/**/*.{js,ts,jsx,tsx}',
  ]
...
};
```

**Comment by gyaaniguy** [comment]
Edit: Figured it out!  Problem was source.css . 
I had put the compiled css file in the source css. Should have been
```
@import "tailwindcss/base";
@import "tailwindcss/components";
@import "tailwindcss/utilities";
```
Would really appreciate to know what I am doing wrong. On Windows. Tried various combinations of 'purge-paths'
Config:
```
module.exports = {
  purge: {
    enabled: true,
    content: ['*.html'],
  },
}
```
Dir Structure:
![image](https://user-images.githubusercontent.com/929990/121658584-1aac5d00-cabf-11eb-9674-e03760f351a7.png)
Command 
```
PS D:\work\code\web\> npx tailwindcss-cli build tailwind.css -o tailwind1.css
   tailwindcss 2.1.4
   🚀 Building: tailwind.css
   ✅ Finished in 3 s
   📦 Size: 3.81MB
   💾 Saved to tailwind1.css
```

HTML:
```
<!doctype html>
<head>
    <link href="tailwind.css" rel="stylesheet">
    <title>ty</title>
</head>
<body>   <div class="pt-6 space-y-4">     </div> </body>
</html>
```
